clients/src/main/java/org/apache/kafka/common/Cluster.java

```diff
@@ -260,16 +225,28 @@ public Integer partitionCountForTopic(String topic) {
      * @return A list of partitions
      */
     public List<PartitionInfo> partitionsForNode(int nodeId) {
-        List<PartitionInfo> parts = this.partitionsByNode.get(nodeId);
-        return (parts == null) ? Collections.emptyList() : parts;
+        return partitionsByTopicPartition.values().stream()
+                .filter(info -> Objects.nonNull(info.leader()) && info.leader().id() == nodeId)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Remove a partition from this cache.
+     * @param topicPartition
+     * @return
+     */
+    public boolean removePartition(TopicPartition topicPartition) {
```

Comment:
This is still a public API, so we would need a KIP for this change. That said, with the nice simplifications here, it might be reasonable to replace usage of the public `Cluster` object with a similar internal object. This would reduce the temptation to leak internal details into the public API.

Reply:
In that case, could we make Cluster an interface and introduce a MutableCluster interface which is used internally? The implementation currently in Cluster would move a new and "internal" DefaultCluster class and implement both Cluster and MutableCluster. `Metadata#fetch` would return the Cluster interface, but within Metadata when we need to update/delete stuff we can use the methods from MutableCluster. Seem reasonable?

Or, like you said, we could do a KIP for making Cluster mutable.